feign-opentracing/src/main/java/feign/opentracing/HttpHeadersInjectAdapter.java

```diff
@@ -32,7 +35,23 @@ public void put(String key, String value) {
             headers.put(key, values);
         }
 
-        values.add(value);
+        try {
+            values.add(value);
+        } catch (UnsupportedOperationException ex) {
+            if (values instanceof List) {
+                // Handle unmodifiable Lists
+                List<String> list = new ArrayList<>(values);
+                list.add(value);
+                headers.put(key, list);
+            } else if (values instanceof Set) {
```

Comment:
Why do we have to distinguish between the list and set?

Reply:
Since the function is generic and can receive both Lists and Sets, I think it would be a bad idea to change the type of the header collection. We don't know what header processing there is later in the user's app.